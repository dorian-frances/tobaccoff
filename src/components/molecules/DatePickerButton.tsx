import { Text, TouchableRipple } from 'react-native-paper';
import { Platform, View } from 'react-native';
import Icon from '@expo/vector-icons/MaterialCommunityIcons';
import React, { useState } from 'react';
import { Colors } from '../../assets/colors/colors.enum';
import DateTimePicker from '@react-native-community/datetimepicker';

type DatePickerButtonProps = {
  infoText: string;
};

const DatePickerButton = ({
  infoText = 'Lorem Ipsum',
}: DatePickerButtonProps) => {
  const [date, setDate] = useState(new Date());
  const [mode, setMode] = useState('date');
  const [show, setShow] = useState(false);

  const onChange = (event: any, selectedDate: any) => {
    setShow(false);
    setDate(selectedDate);
  };

  const showMode = (mode: string) => {
    if (Platform.OS === 'android') {
      setShow(false);
    }
    setShow(true);
    setMode(mode);
  };
  const showDatePicker = () => {
    showMode('date');
  };
  return (
    <View>
      <TouchableRipple
        borderless={true}
        onPress={showDatePicker}
        style={{
          borderRadius: 100,
          height: 50,
        }}
      >
        <View
          style={{
            flexDirection: 'row',
            alignItems: 'center',
            borderColor: 'black',
            borderWidth: 1,
            borderRadius: 100,
            height: 50,
            backgroundColor: Colors.INPUT_BACKGROUND_COLOR,
          }}
        >
          <View style={{ marginLeft: 15 }}>
            <Icon name={'calendar-blank'} size={25} />
          </View>
          <View style={{ marginLeft: 15 }}>
            <Text style={{ color: Colors.INDICATIVE_TEXT_COLOR }}>
              {infoText}
            </Text>
            <Text style={{ color: Colors.BUSINESS_TEXT_COLOR }}>
              {date.toLocaleDateString()}
            </Text>
          </View>
        </View>
      </TouchableRipple>
      <View>
        {show && (
          <DateTimePicker
            testID="dateTimePicker"
            value={date}
            is24Hour={true}
            onChange={onChange}
          />
        )}
      </View>
    </View>
  );
};

export default DatePickerButton;
