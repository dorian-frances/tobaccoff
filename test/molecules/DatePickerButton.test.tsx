import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import DatePickerButton from '../../src/components/molecules/DatePickerButton';

describe('<DatePickerButton/>', () => {
  const infoText = 'fakeInfoText';
  const fakeSetStopDate = jest.fn();

  it('should render correctly', () => {
    const component = render(
      <DatePickerButton
        infoText={infoText}
        defaultDate={new Date(2023, 2, 18)}
        getStopDate={fakeSetStopDate}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render correctly with default date set to now', () => {
    const defaultDate = new Date(2023, 2, 18);
    const component = render(
      <DatePickerButton
        infoText={infoText}
        defaultDate={defaultDate}
        getStopDate={fakeSetStopDate}
      />
    );
    const dateText = component.getByTestId('dateText');
    expect(dateText.props.children).toEqual(defaultDate.toLocaleDateString());
  });

  it('should not show dateTimePicker when initialising', () => {
    const component = render(
      <DatePickerButton
        infoText={infoText}
        defaultDate={new Date(2023, 2, 18)}
        getStopDate={fakeSetStopDate}
      />
    );
    expect(() => component.getByTestId('dateTimePicker')).toThrowError();
  });

  it('should show dateTimePicker on click', () => {
    const component = render(
      <DatePickerButton
        infoText={infoText}
        defaultDate={new Date(2023, 2, 18)}
        getStopDate={fakeSetStopDate}
      />
    );
    const touchableView = component.getByTestId('dateTimePickerTouchable');
    fireEvent.press(touchableView);
    const dateTimePicker = component.getByTestId('dateTimePicker');
    expect(dateTimePicker).toBeTruthy();
  });

  it('should change value on selected dated', async () => {
    const component = render(
      <DatePickerButton
        infoText={infoText}
        defaultDate={new Date(2023, 2, 18)}
        getStopDate={fakeSetStopDate}
      />
    );

    const touchableView = await component.getByTestId(
      'dateTimePickerTouchable'
    );
    fireEvent.press(touchableView);

    const newDate = new Date(2023, 0, 1);
    const dateTimePicker = await component.getByTestId('dateTimePicker');
    fireEvent(dateTimePicker, 'onChange', {
      nativeEvent: {
        timestamp: newDate,
      },
    });
    const displayedDate = await component.getByTestId('dateText');
    expect(displayedDate.props.children).toEqual(newDate.toLocaleDateString());
  });
});
