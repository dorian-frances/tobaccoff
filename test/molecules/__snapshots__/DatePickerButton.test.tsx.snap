// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePickerButton/> should render correctly 1`] = `
<View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "overflow": "hidden",
        },
        false,
        {
          "borderRadius": 100,
          "height": 50,
        },
      ]
    }
    testID="dateTimePickerTouchable"
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#79747E",
          "borderRadius": 100,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 50,
        }
      }
    >
      <View
        style={
          {
            "marginLeft": 15,
          }
        }
      >
        <Text />
      </View>
      <View
        style={
          {
            "marginLeft": 15,
          }
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "color": "#79747E",
              },
            ]
          }
        >
          fakeInfoText
        </Text>
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "color": "#000000",
              },
            ]
          }
          testID="dateText"
        >
          18/03/2023
        </Text>
      </View>
    </View>
  </View>
  <View />
</View>
`;
